site/content/blog/2021-04-23-optimizing-go-compiler-2.md

```diff
@@ -0,0 +1,320 @@
+
+-   `MOVXstore` stores the contents from the second argument into the first one, with the amount of bytes moved depending on the prefix `X`
+-   `SHRQconst` shifts the number in its only argument a number of places specified by the auxiliary argument, the number in square brackets
+
+Let's see the lines again, one by one. The first one was:
+
+```
+v171 (+84) = MOVBstore <mem> v248 v34 v1
+```
+
+This one is straightforward: it stores a single **B**yte from `v34` (the contents in `temp`) into the memory pointed to by `v248` (which is `dest`). That is, we are copying `temp[0]` into `dest[0]`.
+
+The second one is more interesting, let's break it down:
+
+```
+                   this L means 4 bytes        ┏ dest               temp ┓
+                   ━━━━━━━━┳━━━━━━━━━━━        ┃                         ┃
+                           ┃                ━━━┛                         ┗━━
+     new                                                                               previous
+state of ━┫ v219 (88) = MOVLstore <mem> [1] v248 (SHRQconst <uint64> [8] v34) v171 ┣━━ state of
+  memory                                                                               memory
+                                        ━┳━                          ━┳━
+                                         ┃                            ┃
+         ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓      ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
+           auxiliary argument to MOVLstore, so it           auxiliary argument to SHRQconst, so
+           starts writing data one byte after the           it shifts the contents 8 bits to the
+           memory pointed to by the second argument         right, effectively discarding them
+         ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛      ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
+```
+
+Now we're storing a **L**ong word (4 bytes) from the second argument, `(SHRQconst <uint64> [8] v34)`, into the memory pointed to by `v248`, which is `dest`. But we have the `[1]` auxiliary argument! So we don't write directly into the memory pointed to by `v248`, but one byte after that pointer: that is, the first byte of `dest` is left untouched, and we write contents from the second byte on. In terms of Go code, as we are storing 4 bytes, it means that we're writing to `dest[1]`, `dest[2]`, `dest[3]` and `dest[4]`.
+
+But what 4 bytes do we write? That's what `(SHRQconst <uint64> [8] v34)` is telling us: we already wrote `temp[0]` in `dest[0]`, so it would make sense to read now from `temp[1]` on. And that's exactly what we are doing with the shift: as the auxiliary argument is `[8]`, that means that we are shifting the content of `v34` (or `temp`) to the right 8 bits. Those 8 bits are already in `dest[0]`, so we can safely discard them and only read the next ones: from `temp[1]` on. As `MOVLstore` is writing 4 bytes, that means that we are getting 4 bytes worth of data; that is: `temp[1]`, `temp[2]`, `temp[3]` and `temp[4]`.
```

Comment:
```suggestion
But what 4 bytes do we write? That's what `(SHRQconst <uint64> [8] v34)` is telling us: We already wrote `temp[0]` in `dest[0]`, so it would make sense to read now from `temp[1]` on. And that's exactly what we're doing with the shift: As the auxiliary argument is `[8]`, that means that we're shifting the content of `v34` (or `temp`) to the right 8 bits. Those 8 bits are already in `dest[0]`, so we can safely discard them and only read the next ones: from `temp[1]` on. As `MOVLstore` is writing 4 bytes, that means that we're getting 4 bytes' worth of data; that is: `temp[1]`, `temp[2]`, `temp[3]`, and `temp[4]`.
```